contracts/AINFT721.sol

```diff
@@ -0,0 +1,304 @@
+    using Strings for uint256;
+
+    struct MetadataContainer {
+        address updater;
+        string metadataURI;
+    }
+    bool public immutable IS_CLONED;
+    IERC721 public immutable ORIGIN_NFT;
+    address public PAYMENT_PLUGIN;
+    bytes32 public constant PAUSER_ROLE = keccak256("PAUSER_ROLE");
+    bytes32 public constant MINTER_ROLE = keccak256("MINTER_ROLE");
+
+    Counters.Counter private _tokenIdCounter;
+    string private baseURI;
+    mapping(bytes32 => MetadataContainer) private _metadataStorage; // keccak256(bytes(tokenId, <DELIMETER>, version)) => MetadataContainer
+    mapping(uint256 => uint256) private _tokenURICurrentVersion; // tokenId => tokenURIVersion
+
+    constructor(string memory name_, string memory symbol_, bool isCloned_, address originNFT_) ERC721(name_, symbol_) {
+        if (isCloned_) {
+            //FIXME: need to use tx.origin?
+            _grantRole(DEFAULT_ADMIN_ROLE, tx.origin);
+            _grantRole(PAUSER_ROLE, tx.origin);
+            _grantRole(MINTER_ROLE, tx.origin);
+
+            
+        } else {
+            //FIXME: the originNFT contract address should not be ROLED
+            _grantRole(DEFAULT_ADMIN_ROLE, tx.origin);
+            _grantRole(PAUSER_ROLE, tx.origin);
+            _grantRole(MINTER_ROLE, tx.origin);
+        }
+        ORIGIN_NFT = IERC721(originNFT_);
+        IS_CLONED = isCloned_;
+        require((address(ORIGIN_NFT) == address(0) && !IS_CLONED) ||
+                (address(ORIGIN_NFT) != address(0) && IS_CLONED), 
+                "If AINFT721 is created first time, the originNFT_ should not be SET. \
+                \n If AINFT721 is cloned by something, the originNFT_ should be SET.");
+    }
+
+    modifier Cloned () {
+        require(IS_CLONED, "Only cloned contract can execute.");
+        _;
+    }
+    
+    function mintFromOriginInstance(
+        uint256 tokenId_
+    ) public Cloned {
+        require(!_exists(tokenId_), "The tokenId_ is already minted or cloned");
+        address originOwner = ORIGIN_NFT.ownerOf(tokenId_);
+        require(_msgSender() == originOwner, "The sender should be the holder of origin NFT.");
+        _safeMint(_msgSender(), tokenId_);
+
+    }
+    
+    function mintFromOriginInstanceOnBehalf(
+        uint256[] calldata tokenIds_,
+        address[] calldata recipients_
+    ) public Cloned onlyRole(MINTER_ROLE) {
+        for (uint i = 0; i < tokenIds_.length; i++) {
+            require(!_exists(tokenIds_[i]), "The tokenId_ is already minted or cloned");
+            require(recipients_[i] == ORIGIN_NFT.ownerOf(tokenIds_[i]), "The sender should be the holder of origin NFT.");
+            _safeMint(recipients_[i], tokenIds_[i]);
+        }
+    }
+
+    function setPaymentContract(address paymentPlugin_) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(_msgSender() == tx.origin && _msgSender() != address(0), "Only EOA can set payment contract.");
+        PAYMENT_PLUGIN = paymentPlugin_;
```

Comment:
I think we need to check the `paymentPlugin_` is already pointed to this contract. (because of the callflows)
- AINFT721 created
- AINPayment created (with these AINFT721 address)
- AINFT721 setPaymentContract (with above)
  - so we can check this in here.

Reply:
Is there any desirable way that `setPaymentContract` can check whether `AINPayment` is already pointed to this contract?

The first method I've thought was to make `connect` function inside `AINPayment`, which calls `AINFT721::setPaymentContract()`, but it occurs vulnerability that anyone who deploys `AINPayment` can change the `PAYMENT_PLUGIN`.

Reply:
I'm going to just add  `require(paymentPlugin_.isContract());` to check whether `paymentPlugin_` is contract address or not. Any good ways are welcomed!

Reply:
In my first glance, I was thinking the way by checking paymentPlugin_.ainft value is same with this contract's address. just value comparision. that's all.

Reply:
It would be discussed on #4 . Thanks!